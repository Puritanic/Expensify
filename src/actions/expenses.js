import database from '../firebase/firebase';

// * Actions workflow
// ? component calls action generator
// ? action generator returns function
// ? component dispatches function (redux middleware)(redux-thunk?)
// ? redux store changes

// ADD_EXPENSE
export const addExpense = expense => ({
    type: 'ADD_EXPENSE',
    expense,
});

export const startAddExpense = (expenseData = {}) => (dispatch) => {
    // set defaults and destructure from expenseData
    const {
        description = '',
        notes = '',
        amount = 0,
        createdAt = 0,
    } = expenseData;
    const expense = {
        description,
        notes,
        amount,
        createdAt,
    };
    return database
        .ref('expenses')
        .push(expense)
        .then((ref) => {
            dispatch(addExpense({
                id: ref.key,
                ...expense,
            }), );
        });
};
// REMOVE_EXPENSE
export const removeExpense = ({ id } = {}) => ({
    type: 'REMOVE_EXPENSE',
    id,
});
// EDIT_EXPENSE
export const editExpense = (id, updates) => ({
    type: 'EDIT_EXPENSE',
    id,
    updates,
});
