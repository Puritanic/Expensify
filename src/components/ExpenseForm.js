import React, { Component } from 'react';
import moment from 'moment';
import 'react-dates/initialize'; // NOTE: React-dates not working without initializing them first
import { SingleDatePicker } from 'react-dates';
// import { formatPr ice } from '../helpers';

const now = moment();
// console.log(now.format('MMM Do, Y'));

export default class ExpenseForm extends Component {
    constructor(props) {
        super(props);

        this.state = {
            description: props.expense ? props.expense.description : '',
            amount: props.expense
                ? (props.expense.amount / 100).toString()
                : '',
            notes: props.expense ? props.expense.notes : '',
            createdAt: props.expense
                ? moment(props.expense.createdAt)
                : moment(),
            dateFocused: false,
            error: '',
        };
    }

    onDescriptionChange = (e) => {
        const description = e.target.value;
        this.setState(() => ({ description }));
    };
    onNotesChange = (e) => {
        const notes = e.target.value;
        this.setState(() => ({ notes }));
    };

    onAmountChange = (e) => {
        const amount = e.target.value;
        // Regex to format price to max 2 decimal points *
        if (amount.match(/^\d*(\.\d{0,2})?$/)) {
            this.setState(() => ({ amount }));
        }
    };

    onDateChange = (createdAt) => {
        if (createdAt) {
            this.setState(() => ({ createdAt }));
        }
    };

    onFocusChange = ({ focused }) => {
        this.setState(() => ({ dateFocused: focused }));
    };

    onSubmit = (e) => {
        e.preventDefault();
        if (!this.state.description || !this.state.amount) {
            // Set error state
            const error = 'Please provide description and amount.';
            this.setState((state, props) => ({ error }));
        } else {
            // Clear the error
            this.setState(() => ({ error: '' }));

            this.props.onSubmit({
                description: this.state.description,
                amount: parseFloat(this.state.amount, 10) * 100,
                notes: this.state.notes,
                createdAt: this.state.createdAt.valueOf(),
            });
        }
    };

    render() {
        return (
            <div>
                {this.state.error && <p>{this.state.error}</p>}
                <form onSubmit={this.onSubmit}>
                    <input
                        type="text"
                        placeholder="Description"
                        value={this.state.description}
                        onChange={this.onDescriptionChange}
                        required
                    />
                    <input
                        type="text"
                        placeholder="Amount"
                        value={this.state.amount}
                        onChange={this.onAmountChange}
                        required
                    />
                    <SingleDatePicker
                        date={this.state.createdAt} // momentPropTypes.momentObj or null
                        onDateChange={this.onDateChange} // PropTypes.func.isRequired
                        focused={this.state.dateFocused} // PropTypes.bool
                        onFocusChange={this.onFocusChange} // PropTypes.func.isRequired
                        numberOfMonths={1}
                        isOutsideRange={() => false}
                    />
                    <textarea
                        name="notes"
                        cols="30"
                        rows="10"
                        value={this.state.notes}
                        onChange={this.onNotesChange}
                        placeholder="Expense Notes"
                    />
                    <button type="submit">Add Expense</button>
                </form>
            </div>
        );
    }
}

/**

/^\d*(\.\d{0,2})?$/

^ asserts position at start of the string
\d* matches a digit (equal to [0-9])
* Quantifier — Matches between zero and unlimited times, as many times as possible, giving back as needed (greedy)
1st Capturing Group (\.\d{0,2})?
? Quantifier — Matches between zero and one times, as many times as possible, giving back as needed (greedy)
\. matches the character . literally (case sensitive)
\d{0,2} matches a digit (equal to [0-9])
{0,2} Quantifier — Matches between 0 and 2 times, as many times as possible, giving back as needed (greedy)
$ asserts position at the end of the string, or before the line terminator right at the end of the string (if any)

 */
