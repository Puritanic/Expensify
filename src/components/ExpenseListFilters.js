import 'react-dates/initialize';
import React, { Component } from 'react';
import { DateRangePicker } from 'react-dates';
import { connect } from 'react-redux';
import {
  setTextFilter,
  sortByDate,
  sortByAmount,
  setStartDate,
  setEndDate,
} from '../actions/filters';

export class ExpenseListFilters extends Component {
  constructor(props) {
    super();
    this.state = {
      focusedInput: null,
    };
  }
  onDatesChange = ({ startDate, endDate }) => {
    this.props.setStartDate(startDate);
    this.props.setEndDate(endDate);
  };

  onFocusChange = (focusedInput) => {
    this.setState(() => ({ focusedInput }));
  };

  onTextChange = (e) => {
    this.props.setTextFilter(e.target.value);
  };

  onSortChange = (e) => {
    if (e.target.value === 'date') {
      this.props.sortByDate();
    } else if (e.target.value === 'amount') {
      this.props.sortByAmount();
    }
  };

  render() {
    return (
      <div className="content-container">
        <div className="input-group">
          <div className="input-group__item">
            {' '}
            <input
              type="text"
              name="filter"
              value={this.props.filters.text}
              onChange={this.onTextChange}
              placeholder="Search by expense"
            />
          </div>
          <div className="input-group__item">
            {' '}
            <label htmlFor="sortBy">
              Sort by&nbsp;
              <select
                value={this.props.filters.sortBy}
                onChange={this.onSortChange}
                id="sortBy"
              >
                <option value="date">Date</option>
                <option value="amount">Amount</option>
              </select>
            </label>
          </div>
          <div className="input-group__item">
            {' '}
            <DateRangePicker
              startDate={this.props.filters.startDate} // momentPropTypes.momentObj or null,
              endDate={this.props.filters.endDate} // momentPropTypes.momentObj or null,
              startDateId="datepicker_start_home"
              endDateId="datepicker_end_home"
              onDatesChange={this.onDatesChange} // PropTypes.func.isRequired,
              focusedInput={this.state.focusedInput} // PropTypes.oneOf([START_DATE, END_DATE]) or null,
              onFocusChange={this.onFocusChange} // PropTypes.func.isRequired,
              isOutsideRange={() => false}
              numberOfMonths={1}
              showClearDates
            />
          </div>
        </div>
      </div>
    );
  }
}

const mapStateToProps = state => ({
  filters: state.filters,
});

const mapDispatchToProps = dispatch => ({
  setTextFilter: (text) => {
    dispatch(setTextFilter(text));
  },
  sortByDate: () => {
    dispatch(sortByDate());
  },
  sortByAmount: () => {
    dispatch(sortByAmount());
  },
  setStartDate: (startDate) => {
    dispatch(setStartDate(startDate));
  },
  setEndDate: (endDate) => {
    dispatch(setEndDate(endDate));
  },
});

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);
