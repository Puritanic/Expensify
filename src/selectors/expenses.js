import moment from 'moment';

const getVisibleExpenses = (expenses, {
    text, sortBy, startDate, endDate
}) =>
    expenses
        .filter((expense) => {
            const createdAtMoment = moment(expense.createdAt);
            const startDateMatch = startDate
                ? startDate.isSameOrBefore(createdAtMoment, 'day')
                : true;
            const endDateMatch = endDate
                ? endDate.isSameOrAfter(createdAtMoment, 'day')
                : true;
            const textMatch = expense.description
                .toLowerCase()
                .includes(text.toLowerCase());

            return startDateMatch && endDateMatch && textMatch;
        })
        // eslint-disable-next-line array-callback-return
        .sort((a, b) => {
            // from highest to lowest value
            if (sortBy === 'date') {
                return a.createdAt < b.createdAt ? 1 : -1;
            } else if (sortBy === 'amount') {
                return a.amount < b.amount ? 1 : -1;
            }
        });

export default getVisibleExpenses;
