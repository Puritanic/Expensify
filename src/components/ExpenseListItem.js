import React from "react";
import { connect } from "react-redux";
import { removeExpense } from "../actions/expenses";

const ExpenseListItem = ({ description, amount, createdAt, id, dispatch }) => (
    <div>
        <h4>{description}</h4>
        <p>
            {amount} -- {new Date(createdAt * 1000).toDateString()}
            <button
                onClick={() => {
                    dispatch(removeExpense({ id }));
                }}
            >
                &times;
            </button>
        </p>
    </div>
);
/* NOTE: In the first pair of the parenthesis we re getting state info,
    but for this operation we dont need to know anything about state, so we can leave it blank.
*/

export default connect()(ExpenseListItem);
