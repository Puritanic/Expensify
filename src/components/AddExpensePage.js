import React from 'react';
import { connect } from 'react-redux';
import { startAddExpense } from '../actions/expenses';
import ExpenseForm from './ExpenseForm';

export class AddExpensePage extends React.Component {
  onSubmit = (expense) => {
    this.props.startAddExpense(expense);
    this.props.history.push('/');
  };

  render() {
    return (
      <div>
        <div className="content-container">
          <h2>Add expense</h2>
        </div>
        <ExpenseForm onSubmit={this.onSubmit} />
      </div>
    );
  }
}

const mapDispatchToProps = dispatch => ({
  startAddExpense: expense => dispatch(startAddExpense(expense)),
});

export default connect(undefined, mapDispatchToProps)(AddExpensePage);
