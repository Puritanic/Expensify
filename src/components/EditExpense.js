import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { editExpense, removeExpense } from '../actions/expenses';

const EditExpensePage = props => (
    <div>
        <ExpenseForm
            expense={props.expense}
            onSubmit={(expense) => {
                // dispatch the action to edit the expense
                props.dispatch(editExpense(props.expense.id, expense));
                // Redirect to home page
                props.history.push('/');
                console.info('updated', expense);
            }}
        />
        <button
            onClick={() => {
                props.dispatch(removeExpense({ id: props.expense.id }));
                props.history.push('/');
            }}
        >
            Remove Expense
        </button>
    </div>
);

const mapStateToProps = (state, props) => ({
    expense: state.expenses.find(expense => expense.id === props.match.params.id, ),
});

/* NOTE: In the first pair of the parenthesis we re getting state info,
    but if operation doesn't need to know anything about state, we can leave it blank.
*/
export default connect(mapStateToProps)(EditExpensePage);
