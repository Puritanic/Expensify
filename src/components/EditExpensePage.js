import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { startEditExpense, startRemoveExpense } from '../actions/expenses';

export class EditExpensePage extends React.Component {
  onSubmit = (expense) => {
    // dispatch the action to edit the expense
    this.props.startEditExpense(this.props.expense.id, expense);
    // Redirect to home page
    this.props.history.push('/');
    console.info('updated', expense);
  };

  onRemove = () => {
    this.props.startRemoveExpense({ id: this.props.expense.id });
    this.props.history.push('/');
  };

  render() {
    return (
      <div>
        <div className="page-header">
          <div className="content-container">
            <h3 className="page-header__title">Edit Page</h3>
          </div>
        </div>
        <ExpenseForm
          expense={this.props.expense}
          onSubmit={this.onSubmit}
          path={this.props.match.path}
        />
        <div className="content-container--center">
          <button className="button button--remove" onClick={this.onRemove}>
            Remove Expense
          </button>
        </div>
      </div>
    );
  }
}

const mapStateToProps = (state, props) => ({
  expense: state.expenses.find(expense => expense.id === props.match.params.id),
});

const mapDispatchToProps = (dispatch, props) => ({
  startEditExpense: (id, expense) => dispatch(startEditExpense(id, expense)),
  startRemoveExpense: data => dispatch(startRemoveExpense(data)),
});

/* NOTE: In the first pair of the parenthesis we re getting state info,
    but if operation doesn't need to know anything about state, we can leave it blank.
*/
export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);
