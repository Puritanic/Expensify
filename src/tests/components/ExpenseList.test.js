import React from 'react';
import { ExpenseList } from '../../components/ExpenseList';
import expenses from '../fixtures/expenses';
/* eslint-disable no-undef */
it('should render ExpenseList with expenses', () => {
    const wrapper = shallow(<ExpenseList expenses={expenses} />);
    expect(toJson(wrapper)).toMatchSnapshot();
});

it('should render ExpenseList without expenses', () => {
    const wrapper = shallow(<ExpenseList expenses={[]} />);
    expect(toJson(wrapper)).toMatchSnapshot();
});
