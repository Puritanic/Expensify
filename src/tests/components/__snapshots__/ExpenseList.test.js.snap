// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpenseList with expenses 1`] = `
<div>
  <ExpenseListItem
    amount={190}
    createdAt={0}
    description="Gym"
    id="1"
    key="1"
    notes=""
  />
  <ExpenseListItem
    amount={1000}
    createdAt={-345600000}
    description="Cable"
    id="2"
    key="2"
    notes=""
  />
  <ExpenseListItem
    amount={410}
    createdAt={345600000}
    description="Food"
    id="3"
    key="3"
    notes=""
  />
</div>
`;

exports[`should render ExpenseList without expenses 1`] = `
<div>
  <p>
    You have no Expenses
  </p>
</div>
`;
